File: Python/ttio_udp.py
import socket
import time

TTIO_HOST = '104.199.85.211'
TTIO_PORT = 28399

def sendUDPMessageTTIOv0(productId, hashStr, deviceId, messageStr, sha1Str = None):
    # by deviceId
    protocol = '0x00'

    idProductHex = decimalToPaddedHexString(productId)
    idProductLengthHex = decimalToPaddedHexString((len(idProductHex)-2)/2)
    
    hashHex = stringToHex(hashStr)
    
    idDeviceHex = stringToHex(deviceId)
    idDeviceLengthHex = decimalToPaddedHexString((len(idDeviceHex)-2)/2)

    mg = messageStr

    sha1 = stringToHex(sha1Str) if sha1Str is not None else None

    hexList = [protocol, idProductLengthHex, idProductHex, hashHex, idDeviceLengthHex, idDeviceHex, mg, sha1]
    print(hexList)
    message = concatenateListOfHex(hexList)

    sendUDPMessage(message, TTIO_HOST, TTIO_PORT)

def sendUDPMessageTTIOv1(thingToken, messageStr, sha1Str = None):
    # by token
    protocol = '0x01'
    token = stringToHex(thingToken)
    mg = messageStr
    sha1 = stringToHex(sha1Str) if sha1Str is not None else None

    hexList = [protocol, token, mg, sha1]
    print(hexList)
    message = concatenateListOfHex(hexList)

    sendUDPMessage(message, TTIO_HOST, TTIO_PORT)

def sendUDPMessage(message, TTIO_HOST, TTIO_PORT):
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.sendto(bytearray.fromhex(message),(TTIO_HOST,TTIO_PORT))
    print(message)
    print("UDP message sent to '{}':'{}'".format(TTIO_HOST,TTIO_PORT))
    time.sleep(1)
    data, address = s.recvfrom(4096)
    print("received %s from address %s"% (data,address))

def decimalToPaddedHexString(num):
    paddedHex = "{0:#0{1}x}".format(num,4)
    return paddedHex

def concatenateListOfHex(hexList):
    return ''.join([format(int(c, 16), '02X') for c in hexList if c is not None])
    
def stringToHex(str):
    str_hex = list()
    for character in str:
        str_hex.append(hex(ord(character)))
    return '0x' + concatenateListOfHex(str_hex)
    

File: JavaScript/lib/ttio_udp.js
const dgram = require('dgram');

const TTIO_HOST = '104.199.85.211';
const TTIO_PORT = 28399; 

exports.sendUDPMessageTTIOv0 = function(productId, hash, deviceId, messageStr, sha1Str = null){

    // by deviceId
    let protocol = Buffer.from('00', 'hex');
   
    var productIdHex = decimalToPaddedHexString(productId);
    let idProductHex = Buffer.from(productIdHex, 'hex'); 

    var idProductLength = idProductHex.length.toString(16)
    if (idProductLength.length == 1) idProductLength = '0' + idProductLength;
    let idProductLengthHex = Buffer.from(idProductLength, 'hex');

    let hashHex = Buffer.from(hash.toString('hex'));

    let idDeviceHex = Buffer.from(deviceId.toString('hex'));

    var deviceIdLength = deviceId.length.toString(16);
    if (deviceIdLength.length == 1) deviceIdLength = '0' + deviceIdLength;
    var idDeviceLength = idDeviceHex.length.toString(16)
    if (idDeviceLength.length == 1) idDeviceLength = '0' + idDeviceLength;
    let idDeviceLengthHex = Buffer.from(idDeviceLength, 'hex');

    let sha1 = sha1Str != null ? Buffer.from(sha1Str, 'hex') : Buffer.from('', 'hex');

    let message = Buffer.concat([protocol, idProductLengthHex, idProductHex, hashHex, idDeviceLengthHex, idDeviceHex, Buffer.from(messageStr, 'hex'), sha1]);

    sendUDPMessage(message, TTIO_HOST, TTIO_PORT)
}

exports.sendUDPMessageTTIOv1 = function(thingToken, messageStr, sha1Str = null){

    // by token
    let protocol = Buffer.from('01', 'hex');
    let token = Buffer.from(thingToken.toString('hex'));
    let sha1 = sha1Str != null ? Buffer.from(sha1Str, 'hex') : Buffer.from('', 'hex');
    let message = Buffer.concat([protocol, token, Buffer.from(messageStr, 'hex'), sha1]);

    sendUDPMessage(message, TTIO_HOST, TTIO_PORT)
}

function sendUDPMessage(message, TTIO_HOST, TTIO_PORT){
    var client = dgram.createSocket('udp4');

    client.on('error', function(e) {
        throw e;
    });

    client.on('message', function (msg, rinfo) {
        console.log(msg);
        client.close();
    });

    client.send(message, 0, message.length, TTIO_PORT, TTIO_HOST, function(err, bytes) {
        console.log(message);
        if (err) throw err;
        console.log('UDP message sent to ' + TTIO_HOST +':'+ TTIO_PORT);
    });
}

function decimalToPaddedHexString(num)
{
    var hex = num.toString(16)
    if (hex.length % 2 != 0) var paddedHex = '0' + hex;
    else paddedHex = hex;
    return paddedHex
}